src/integrationtests/java/com/aws/greengrass/integrationtests/e2e/mqttclient/MqttTest.java

```diff
@@ -59,10 +61,13 @@ void GIVEN_mqttclient_WHEN_subscribe_and_publish_THEN_receives_all_messages()
         client.subscribe(SubscribeRequest.builder().topic("A/B/C").callback((m) -> {
             cdl.countDown();
             logger.atInfo().kv("remaining", cdl.getCount()).log("Received 1 message from cloud.");
+            Integer id = Integer.parseInt(new String(m.getPayload(), StandardCharsets.UTF_8));
+            assertEquals(id, ++lastMessageId);
```

Comment:
lastMessageId should be an AtomicInteger, otherwise this method can have multi threading problems.

Reply:
I am planing the second message is sent out after the first QOS 1 message has been successfully published and should not have any race here. If it exist, that would be the problem of the future from publish method.